code/datums/ai_controllers/_ai_controller.dm

```diff
@@ -0,0 +1,338 @@
+	///AI paused time
+	var/paused_until = 0
+
+/datum/ai_controller/New(atom/new_pawn)
+	change_ai_movement_type(ai_movement)
+	init_subtrees()
+
+	if(idle_behavior)
+		idle_behavior = new idle_behavior()
+
+	PossessPawn(new_pawn)
+
+/datum/ai_controller/Destroy(force, ...)
+	set_ai_status(AI_STATUS_OFF)
+	UnpossessPawn(FALSE)
+	return ..()
+
+///Sets the current movement target, with an optional param to override the movement behavior
+/datum/ai_controller/proc/set_movement_target(source, atom/target, datum/ai_movement/new_movement)
+	movement_target_source = source
+	current_movement_target = target
+	if(new_movement)
+		change_ai_movement_type(new_movement)
+
+///Overrides the current ai_movement of this controller with a new one
+/datum/ai_controller/proc/change_ai_movement_type(datum/ai_movement/new_movement)
+	ai_movement = SSai_movement.movement_types[new_movement]
+
+///Completely replaces the planning_subtrees with a new set based on argument provided, list provided must contain specifically typepaths
+/datum/ai_controller/proc/replace_planning_subtrees(list/typepaths_of_new_subtrees)
+	planning_subtrees = typepaths_of_new_subtrees
+	init_subtrees()
+
+///Loops over the subtrees in planning_subtrees and looks at the ai_controllers to grab a reference, ENSURE planning_subtrees ARE TYPEPATHS AND NOT INSTANCES/REFERENCES BEFORE EXECUTING THIS
+/datum/ai_controller/proc/init_subtrees()
+	if(!LAZYLEN(planning_subtrees))
+		return
+	var/list/temp_subtree_list = list()
+	for(var/subtree in planning_subtrees)
+		var/subtree_instance = SSai_controllers.ai_subtrees[subtree]
+		temp_subtree_list += subtree_instance
+	planning_subtrees = temp_subtree_list
+
+///Proc to move from one pawn to another, this will destroy the target's existing controller.
+/datum/ai_controller/proc/PossessPawn(atom/new_pawn)
+	if(pawn) //Reset any old signals
+		UnpossessPawn(FALSE)
+
+	if(istype(new_pawn.ai_controller)) //Existing AI, kill it.
+		QDEL_NULL(new_pawn.ai_controller)
+
+	if(TryPossessPawn(new_pawn) & AI_CONTROLLER_INCOMPATIBLE)
+		qdel(src)
+		CRASH("[src] attached to [new_pawn] but these are not compatible!")
+
+	pawn = new_pawn
+	pawn.ai_controller = src
+
+	SEND_SIGNAL(src, COMSIG_AI_CONTROLLER_POSSESSED_PAWN)
+
+	reset_ai_status()
+	RegisterSignal(pawn, COMSIG_MOB_STATCHANGE, PROC_REF(on_stat_changed))
+	RegisterSignal(pawn, COMSIG_MOB_LOGIN, PROC_REF(on_sentience_gained))
+
+/// Sets the AI on or off based on current conditions, call to reset after you've manually disabled it somewhere
+/datum/ai_controller/proc/reset_ai_status()
+	set_ai_status(get_expected_ai_status())
+
+/// Returns what the AI status should be based on current conditions.
+/datum/ai_controller/proc/get_expected_ai_status()
+	var/final_status = AI_STATUS_ON
+
+	if (!ismob(pawn))
+		return final_status
+
+	var/mob/living/mob_pawn = pawn
+
+	if(!continue_processing_when_client && mob_pawn.client)
```

Comment:
I'm on mobile but make sure you've also got the new update z level mob procs for this

Reply:
thanks b